### PR TITLE
Add support for enable_confidential_storage flag

### PR DESCRIPTION
**Pull request by Sneha-at**
Adds support for new parameter "enable_confidential_storage" while creating cluster and node-pool. This flag does not support update operations.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
container: added `enable_confidential_storage` to `node_config` in `google_container_cluster` and `google_container_node_pool` (beta)
```

### CODE REVIEW

**Comment by shuyama1** [comment]
@Sneha-at thanks for making the change. Would you mind rebasing your PR to resolve the conflicts? Plus, please check the internal contribution page on how to get the tests automatically run for your PR so that you'll be able to see test results faster

**Comment by sunnylovestiramisu** [comment]
/hold

Until prod fix is out, but please review the change ahead to save some time.

File: mmv1/third_party/terraform/services/container/resource_container_cluster_test.go.erb
Comment (shuyama1) [inline mmv1/third_party/terraform/services/container/resource_container_cluster_test.go.erb:8620]: only nit: should be spaces instead of tabs here for spacing

File: mmv1/third_party/terraform/services/container/resource_container_node_pool_test.go.erb
Comment (shuyama1) [inline mmv1/third_party/terraform/services/container/resource_container_node_pool_test.go.erb:3642]: nit: spcing

**Review by shuyama1 — COMMENTED** [review]
Code looks good to me. Please add a test coverage on negative value `enable_confidential_storage = false`. Just want to make sure that the API will send the value back if we explicitly send it, otherwise it will cause a drift in TF. Thanks

**Comment by shuyama1** [comment]
@Sneha-at I see this PR is still in draft mode. Let me know when it's ready to review or if you need me to trigger build and tests for it. Also please see https://github.com/GoogleCloudPlatform/magic-modules/pull/9122#issuecomment-1739913980 to get your GitHub account join in the right org so the tests will automatically triggered for your PR.

**Comment by Sneha-at** [comment]
Thanks @shuyama1 I have joined "GoogleCloudPlatform" org, how do I trigger the tests?

**Comment by Sneha-at** [comment]
I have a test which uses this flag in node_config (not node_pool{node_config}), is this deprecated now? Or is is necessary to have test with "primary" tag -- `"google_container_cluster""primary"`?
```
func testAccContainerCluster_withConfidentialBootDisk(clusterName, kmsKeyName string) string {
	return fmt.Sprintf(`
resource "google_container_cluster" "with_confidential_boot_disk" {
.....
  	node_config {
	enable_confidential_storage = true
 }
}
```

**Comment by shuyama1** [comment]
@Sneha-at I think the tests will automatically triggered when there's a new action to the PR, like new commits or merge/rebase of main. You don't need to manually comment the command :) Plus, I think the comment for rerun probably only works for the maintainers of the repo

**Comment by Sneha-at** [comment]
Ahh, thanks!! Can you please help me understand whats missing in the tests, more in [comment](https://github.com/GoogleCloudPlatform/magic-modules/pull/9122#issuecomment-1745782305)

**Comment by shuyama1** [comment]
> I have a test which uses this flag in node_config (not node_pool{node_config}), is this deprecated now? Or is is necessary to have test with "primary" tag -- `"google_container_cluster""primary"`?
> 
> ```
> func testAccContainerCluster_withConfidentialBootDisk(clusterName, kmsKeyName string) string {
> 	return fmt.Sprintf(`
> resource "google_container_cluster" "with_confidential_boot_disk" {
> .....
>   	node_config {
> 	enable_confidential_storage = true
>  }
> }
> ```

Sorry, haven't got time to do a full review, by only a quick glance though, I think what you read is correct, the missing test detector is complaining that there's no test coverage on the field `node_pool.node_config.enable_confidential_storage` in the `google_container_cluster` resource. I don't see that we have deprecated `node_pool`, also I couldn't find the deprecation massage on the API page https://cloud.google.com/kubernetes-engine/docs/reference/rest/v1beta1/projects.locations.clusters, but I am super familiar with the resource, is `node_pool.node_config.enable_confidential_storage` a valid resource configuration?

**Comment by Sneha-at** [comment]
Yes `node_pool.node_config.enable_confidential_storage` is a valid resource. I see we use node_config only to set node_defaults in the cluster. So moving this to `node_pool.node_config` in tests.

**Comment by Sneha-at** [comment]
Looks like we need to have test cases for both ways to set the node_config

**Comment by Sneha-at** [comment]
The support for machine type mentioned in tests is getting rolled out and would take some time to be available. The presubmits should pass once we have that ready. Meanwhile, please let me know if everything else looks good.

**Comment by Sneha-at** [comment]
Moving it to "review" to allow presubmits.

**Comment by Sneha-at** [comment]
I wonder why only the Cluster tests are failing while the node_pool tests are passing, is there something else needed for those tests?

**Comment by Sneha-at** [comment]
I don't see any tests for node_pool "TestAccContainerNodePool" in [VCR-test details](https://pantheon.corp.google.com/cloud-build/builds;region=global/ef683f66-21c8-4cbd-aa47-eaf1b273d62b;step=20?e=-13802955&mods=logs_tg_prod&project=graphite-docker-images) last step so assuming it is intended

**Comment by Sneha-at** [comment]
@shuyama1  All the tests are failing due to `deletion_protection` mismatch, are we expected to set this to `true` or `false`? I would want the resources to be cleaned up post execution.
Do I add `ImportStateVerifyIgnore: []string{"deletion_protection"}` ?
```
ImportStateVerify attributes not equivalent. Difference is shown below. Top is actual, bottom is expected.
        
        (map[string]string) (len=1) {
         (string) (len=19) "deletion_protection": (string) (len=4) "true"
        }  
        (map[string]string) (len=1) {
         (string) (len=19) "deletion_protection": (string) (len=5) "false"
```

**Comment by sunnylovestiramisu** [comment]
The feature enablement on the GKE side is completed, the tests should pass now. 

/gcbrun

**Comment by Sneha-at** [comment]
The tests are failing due to `deletion_protection` flag. @shuyama1 could you please help me with what changes are needed here? I added `deletion_protection: false` as we want to clean up clusters post test completion. But this is failing with error, I also tried adding `ImportStateVerifyIgnore: []string{"deletion_protection"},`but still got the same error
```
ImportStateVerify attributes not equivalent. Difference is shown below. Top is actual, bottom is expected.
        
        (map[string]string) (len=1) {
         (string) (len=19) "deletion_protection": (string) (len=4) "true"
        }
        (map[string]string) (len=1) {
         (string) (len=19) "deletion_protection": (string) (len=5) "false"
        }
```

**Comment by shuyama1** [comment]
@Sneha-at I believe adding `deletion_protection: false` to the config and `deletion_protection` to `ImportStateVerifyIgnore` should resolve the issue. Would you mind adding those in your PR so I can take a look at the debug log see what the issue might be? Thanks!

**Comment by Sneha-at** [comment]
The tests seems to be failing with error 
```
Error: Post "https://container.googleapis.com/v1beta1/projects/ci-test-project-188019/locations/us-central1-a/clusters?alt=json&prettyPrint=false": Requested interaction not found
```
Not sure if this is related to my changes or is an infra issue.

**Comment by Sneha-at** [comment]
All the tests have passed during "Recording" mode but 2 tests failed during "Replaying" mode, do we have to worry about "Replaying" mode or are these failures temporary?

**Comment by shuyama1** [comment]
@Sneha-at Tests passed in recording but failed after in replaying usually due to non-determinism or randomness when the VCR replayed the response after the HTTP request was made. We should also get the tests passed in replaying or skip them in VCR if they're not suitable for VCR. I am looking into the failures and will suggest the next step after I figured out why it failed in replaying.

File: mmv1/third_party/terraform/services/container/resource_container_node_pool_test.go.erb
Comment (shuyama1) [inline mmv1/third_party/terraform/services/container/resource_container_node_pool_test.go.erb:3706]: nit: spaces

File: mmv1/third_party/terraform/services/container/resource_container_cluster_test.go.erb
Comment (Sneha-at) [reply mmv1/third_party/terraform/services/container/resource_container_cluster_test.go.erb]: Updated all resource configurations. Please let me know if I missed any

**Review by shuyama1 — CHANGES_REQUESTED** [review]
Sorry forgot that we still need to manually add documentation for handwritten resources. https://googlecloudplatform.github.io/magic-modules/develop/resource/#add-documentation
my apologies!

File: mmv1/third_party/terraform/website/docs/r/container_cluster.html.markdown
Comment (shuyama1) [inline mmv1/third_party/terraform/website/docs/r/container_cluster.html.markdown:782]:
```suggestion
Enabling Confidential Storage will create boot disk with confidential mode. It is disabled by default.
```

**Review by shuyama1 — APPROVED** [review]
Thank you!

**Pull request merged**
